### DISCUSSION TITLE
Sudo prompt not visible

### DISCUSSION THREAD

**Discussion by rrrnld** [kovidgoyal/kitty#5537] [Q&A] [answered]
I have a development build process that I set up like this:

```
ls *.go | entr -rc sudo docker compose up app --build
```

This pipes a list of files to https://github.com/eradman/entr, and whenever they change, runs `sh -c 'sudo docker compose up tax-app --build'`. Because docker needs `sudo`, it will ask me for my password when the last build was a while ago. In other terminal emulators I could see the sudo prompt and enter my password in this case. Kitty shows me an empty screen that I can't exit via `Ctrl + C` or `Ctrl + D`. Similarly when I first start the above process and I would have to enter my credentials for the command to run - the sudo prompt just doesn't show. When closing the terminal window I get a prompt: "Are you sure you want to close this OS window, it has 1 windows running?"

Any idea why, or how to fix it? Thanks for any help!

**Comment by page-down** [comment, marked as answer]
sudo needs tty to get the password.
```
echo test.txt | entr -r tty
not a tty
```

> In other terminal emulators I could see the sudo prompt ...

I can't see the prompt in Konsole.

> ... I can't exit via ... When closing the terminal window I get a prompt ...

That's because `sudo` that can't get the password keeps running in the background. Even if you force close the kitty window, the process will probably still be running.
https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.close_on_child_death

> ... how to fix it? ...

I would recommend not using interactive password authentication on a command that requires repeated automatic execution.

- `docker` itself does not depend on `sudo`, but rather your user is not in the `docker` user group. So you can add your user to the `docker` group and remove `sudo` part.
- Configure`NOPASSWD` in sudoers for the user or specified command.
- Put the entire command in a shell script and use `sudo` to execute the script. (`sudo sh -c 'echo ... | entr ...'`)
- Do not use `entr -r` option (`-r ... Control of the TTY is not transferred to the child process.`)
- Set the `SUDO_ASKPASS` environment variable to point to a GUI askpass program and use the `sudo -A` option. (will pop up a prompt dialog any time needed during monitoring, which is annoying)

**Reply by rrrnld** [reply]
Thanks for the thorough explanation, that really helped me understand what's going on!
